pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +23,102 @@ import (
-// Not safe for concurrent usage
-type KeyGenerator struct {
-	buf []byte
-}
+// hashSetSize is the size selected for hashset used to deduplicate the tags
+// while generating the hash. This size has been selected to have space for
+// approximately 500 tags since it's not impacting much the performances,
+// even if the backend is truncating after 100 tags.
+const hashSetSize = 512
 
 // NewKeyGenerator creates a new key generator
 func NewKeyGenerator() *KeyGenerator {
 	return &KeyGenerator{
-		buf: make([]byte, 0, 1024),
+		seen:  make([]uint64, 512, 512),
+		empty: make([]uint64, 512, 512),
 	}
 }
 
+// KeyGenerator generates hash for the given name, hostname and tags.
+// The tags don't have to be sorted and duplicated tags will be ignored while
+// generating the hash.
+// Not safe for concurrent usage.
+type KeyGenerator struct {
+	// reused buffer to not create a uint64 on the stack every key generation
+	intb uint64
+
+	// seen is used as a hashset to deduplicate the tags when there is more than
+	// 16 and less than 512 tags.
+	seen []uint64
+	// empty is an empty hashset with all values set to 0, to reset `seen`
+	empty []uint64
+
+	// idx is used to deduplicate tags when there is less than 16 tags (faster than the
+	// hashset) or more than 512 tags (hashset has been allocated with 512 values max)
+	idx int
+}
+
 // Generate returns the ContextKey hash for the given parameters.
 // The tags array is sorted in place to avoid heap allocations.
 func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
-	g.buf = g.buf[:0]
+	// between two generations, we have to set the hash to something neutral, let's
+	// use this big value seed from the murmur3 implementations
+	g.intb = 0xc6a4a7935bd1e995
 
-	// Sort the tags in place. For typical tag slices, we use
-	// the in-place insertion sort to avoid heap allocations.
-	// We default to stdlib's sort package for longer slices.
-	// See `pkg/util/sort.go` for info on the threshold.
-	if len(tags) < util.InsertionSortThreshold {
-		util.InsertionSort(tags)
-	} else {
-		sort.Strings(tags)
-	}
+	g.intb = g.intb ^ murmur3.StringSum64(name)
+	g.intb = g.intb ^ murmur3.StringSum64(hostname)
 
-	g.buf = append(g.buf, name...)
-	g.buf = append(g.buf, ',')
-	for i := 0; i < len(tags); i++ {
-		g.buf = append(g.buf, tags[i]...)
-		g.buf = append(g.buf, ',')
+	// there is two implementations used here to deduplicate the tags depending on how
+	// many tags we have to process:
+	//   -  16 < n < hashSetSize:	we use a hashset of `hashSetSize` values.
+	//   -  n < 16 or n > hashSetSize: we use a simple for loops, which is faster than
+	//                         	the hashset when there is less than 16 tags, and
+	//                         	we use it as fallback when there is more than `hashSetSize`
+	//                         	because it is the maximum size the allocated
+	//                         	hashset can handle.
+	if len(tags) > 16 && len(tags) < hashSetSize {
```

Comment:
What do you think about creating 2 functions? One for the case `if len(tags) > 16 && len(tags) < hashSetSize {` and another one for the `else`?

Reply:
Unfortunately, because of the function call, using methods is slightly slower:

```
name                      old time/op  new time/op  delta
KeyGeneration/1-tags-8    22.5ns ± 2%  26.0ns ± 3%  +15.19%  (p=0.008 n=5+5)
KeyGeneration/2-tags-8    31.4ns ± 2%  34.5ns ± 0%  +10.10%  (p=0.008 n=5+5)
KeyGeneration/4-tags-8    47.4ns ± 2%  51.7ns ± 1%   +8.90%  (p=0.008 n=5+5)
KeyGeneration/8-tags-8    89.3ns ± 0%  94.0ns ± 1%   +5.30%  (p=0.008 n=5+5)
KeyGeneration/16-tags-8    199ns ± 0%   211ns ± 3%   +5.94%  (p=0.008 n=5+5)
KeyGeneration/32-tags-8    315ns ± 2%   317ns ± 2%     ~     (p=1.000 n=5+5)
KeyGeneration/64-tags-8    612ns ± 1%   629ns ± 1%   +2.85%  (p=0.008 n=5+5)
KeyGeneration/128-tags-8  1.17µs ± 1%  1.18µs ± 0%   +1.46%  (p=0.008 n=5+5)
```

Reply:
Do you have consistent results? I have run `KeyGeneration` benchmark and have noticed differences when running it several times.

Reply:
Yes. Also, make sure to use `-count` to run each benchmark multiple times.